src/coordinator/storage/local/cluster.go

```diff
@@ -0,0 +1,285 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package local
+
+import (
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/storage"
+	"github.com/m3db/m3db/src/dbnode/client"
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/ident"
+)
+
+var (
+	errNamespaceIDNotSet = errors.New("namespace ID not set")
+	errSessionNotSet     = errors.New("session not set")
+	errRetentionNotSet   = errors.New("retention not set")
+	errResolutionNotSet  = errors.New("resolution not set")
+)
+
+// Clusters is a flattened collection of local storage clusters and namespaces.
+type Clusters interface {
+	io.Closer
+
+	// ClusterNamespaces returns all known cluster namespaces.
+	ClusterNamespaces() []ClusterNamespace
+
+	// UnaggregatedClusterNamespace returns the valid unaggregated
+	// cluster namespace.
+	UnaggregatedClusterNamespace() ClusterNamespace
+
+	// AggregatedClusterNamespace returns an aggregated cluster namespace
+	// at a specific retention and resolution.
+	AggregatedClusterNamespace(attrs RetentionResolution) (ClusterNamespace, bool)
+}
+
+// RetentionResolution is a tuple of retention and resolution that describes
+// an aggregated metrics policy.
+type RetentionResolution struct {
+	Retention  time.Duration
+	Resolution time.Duration
+}
+
+// ClusterNamespace is a local storage cluster namespace.
+type ClusterNamespace interface {
+	NamespaceID() ident.ID
+	Attributes() storage.Attributes
+	Session() client.Session
+}
+
+// UnaggregatedClusterNamespaceDefinition is a definition for an
+// cluster namespace that holds unaggregated metrics data.
+type UnaggregatedClusterNamespaceDefinition struct {
+	NamespaceID ident.ID
+	Session     client.Session
+	Retention   time.Duration
+}
+
+// Validate will validate the cluster namespace definition.
+func (def UnaggregatedClusterNamespaceDefinition) Validate() error {
+	if def.NamespaceID == nil || len(def.NamespaceID.String()) == 0 {
+		return errNamespaceIDNotSet
+	}
+	if def.Session == nil {
+		return errSessionNotSet
+	}
+	if def.Retention <= 0 {
+		return errRetentionNotSet
+	}
+	return nil
+}
+
+// AggregatedClusterNamespaceDefinition is a definition for an
+// cluster namespace that holds aggregated metrics data at a
+// specific retention and resolution.
+type AggregatedClusterNamespaceDefinition struct {
+	NamespaceID ident.ID
+	Session     client.Session
+	Retention   time.Duration
```

Comment:
I'd rather keep it as a separate field, just to make it more similar to `UnaggregatedClusterNamespaceDefinition` struct, just with the addition of the `Resolution` field.  It would be over-de-normalization for this case I believe, I'd prefer to stick with idiomatic.

Reply:
I see what you mean, but figured it might be useful to use it here since you're using RetentionResolution as the key for the deduping map?